File: language-lab/typescript/kernel.ts
import {
  Effect,
  Kind,
  TargetTag,
  cloneFact,
  covers,
  encodeAtom,
  exact,
  frame,
  key,
  materialize,
  offer,
  type Atom,
  type Bytes,
  type Fact,
  type NeedAtom,
  type OfferAtom,
  type Target,
} from "./model.ts";
import { decodeFact, factId } from "./model.ts";

export const Verdict = {
  Valid: "valid",
  Invalid: "invalid",
  Parked: "parked",
  Suppressed: "suppressed",
  Reap: "reap",
  Unknown: "unknown",
} as const;
export type Verdict = (typeof Verdict)[keyof typeof Verdict];

export interface Row {
  readonly owner: Bytes;
  readonly timestamp: bigint;
  readonly atom: Atom;
}

export interface Answer {
  readonly need: NeedAtom;
  readonly rows: readonly Row[];
}

export type Context = readonly Answer[];

export interface Out {
  readonly verdict: Verdict;
  readonly offers: readonly OfferAtom[];
}

export interface Root {
  extract(fact: Fact): boolean;
  project(fact: Fact, context: Context): Out | undefined;
}

export function valid(...offers: readonly OfferAtom[]): Out {
  return { verdict: Verdict.Valid, offers };
}

export function outcome(verdict: Exclude<Verdict, typeof Verdict.Valid>): Out {
  return { verdict, offers: [] };
}

export function by(context: Context, role: Uint8Array): Row[] {
  const wanted = key(role);
  return context.flatMap((answer) => key(answer.need.role) === wanted ? answer.rows : []);
}

function rowKey(row: Row): string {
  return `${key(row.owner)}:${row.timestamp}:${key(encodeAtom(row.atom))}`;
}

function addressKey(role: Uint8Array, scope: Uint8Array): string {
  return key(frame(role, scope));
}

function indexKey(kind: Kind, role: Uint8Array, scope: Uint8Array): string {
  return key(frame(Buffer.from([kind]), role, scope));
}

function targetPoint(target: Target): Bytes | undefined {
  return target.tag === TargetTag.Exact ? target.value : undefined;
}

export class Bucket {
  readonly #exact = new Map<string, { point: Bytes; rows: Row[] }>();
  readonly #ranges: Row[] = [];

  add(row: Row): void {
    const point = targetPoint(row.atom.target);
    if (point !== undefined) {
      const id = key(point);
      const entry = this.#exact.get(id);
      if (entry === undefined) this.#exact.set(id, { point: Buffer.from(point), rows: [row] });
      else entry.rows.push(row);
      return;
    }
    if (row.atom.target.tag === TargetTag.Self) throw new Error("resident SELF target");
    this.#ranges.push(row);
  }

  remove(row: Row): void {
    const point = targetPoint(row.atom.target);
    if (point !== undefined) {
      const id = key(point);
      const entry = this.#exact.get(id);
      if (entry === undefined) return;
      const index = entry.rows.findIndex((candidate) => rowKey(candidate) === rowKey(row));
      if (index >= 0) entry.rows.splice(index, 1);
      if (entry.rows.length === 0) this.#exact.delete(id);
      return;
    }
    const index = this.#ranges.findIndex((candidate) => rowKey(candidate) === rowKey(row));
    if (index >= 0) this.#ranges.splice(index, 1);
  }

  matching(target: Target): Row[] {
    if (target.tag === TargetTag.Self) return [];
    if (target.tag === TargetTag.Exact) {
      const matches = [...(this.#exact.get(key(target.value))?.rows ?? [])];
      for (const row of this.#ranges) if (covers(row.atom.target, target)) matches.push(row);
      return matches;
    }
    const matches: Row[] = [];
    for (const { point, rows } of this.#exact.values()) {
      if (Buffer.compare(target.lo, point) <= 0 && Buffer.compare(point, target.hi) <= 0) {
        matches.push(...rows);
      }
    }
    return matches;
  }

  all(): Row[] {
    return [...this.#exact.values()].flatMap(({ rows }) => rows).concat(this.#ranges);
  }

  get size(): number {
    let result = this.#ranges.length;
    for (const { rows } of this.#exact.values()) result += rows.length;
    return result;
  }
}

const NOW = Buffer.from("now");
const CLOCK = Buffer.from("clock");
const SHIPPED = Buffer.from("shipped");
const WIRE = Buffer.from("wire");

function hostOwner(name: string): Bytes {
  const owner = Buffer.alloc(32);
  owner[0] = 0xff;
  owner.write(name, 1);
  return owner;
}

const NOW_OWNER = hostOwner("now");
const SHIPPED_OWNER = hostOwner("ship");

interface ResidentFact {
  readonly owner: Bytes;
  readonly fact: Fact;
}

export class Node {
  readonly #root: Root;
  readonly #durable = new Map<string, Bytes>();
  readonly #facts = new Map<string, ResidentFact>();
  readonly #rows = new Map<string, Bucket>();
  readonly #memo = new Map<string, Verdict>();
  readonly #clean = new Map<string, Bucket>();
  readonly #owned = new Map<string, Row[]>();
  #frontier: Bytes[] = [];
  #head = 0;
  readonly #queued = new Set<string>();

  constructor(root: Root) {
    this.#root = root;
  }

  admit(data: Uint8Array): Bytes | undefined {
    let fact: Fact;
    try {
      fact = decodeFact(data);
    } catch {
      return undefined;
    }
    const owner = factId(fact);
    const id = key(owner);
    if (this.#facts.has(id)) return Buffer.from(owner);
    this.#facts.set(id, { owner, fact });
    this.#memo.set(id, Verdict.Unknown);
    if (this.#root.extract(cloneFact(fact))) this.#durable.set(id, Buffer.from(data));
    for (const atom of fact.atoms) {
      const bucket = this.#bucket(this.#rows, indexKey(atom.kind, atom.role, atom.scope));
      bucket.add({ owner, timestamp: 0n, atom: materialize(atom, owner) });
    }
    this.#enqueue(owner);
    return Buffer.from(owner);
  }

  offersFor(needAtom: NeedAtom): Row[] {
    return this.#matching(this.#rows.get(indexKey(Kind.Offer, needAtom.role, needAtom.scope)), needAtom.target);
  }

  needsFor(offerAtom: OfferAtom): Row[] {
    return this.#matching(this.#rows.get(indexKey(Kind.Need, offerAtom.role, offerAtom.scope)), offerAtom.target);
  }

  validOffers(needAtom: NeedAtom): Row[] {
    return this.#matching(this.#clean.get(addressKey(needAtom.role, needAtom.scope)), needAtom.target);
  }

  watched(role: Uint8Array, scope: Uint8Array): Row[] {
    return this.#clean.get(addressKey(role, scope))?.all() ?? [];
  }

  turn(nowMs: bigint | undefined, shipped: Iterable<Uint8Array>, bound: number): void {
    if (nowMs !== undefined) {
      if (nowMs < 0n || nowMs > 0xffff_ffff_ffff_ffffn) throw new RangeError("bad clock value");
      const encoded = Buffer.allocUnsafe(8);
      encoded.writeBigUInt64BE(nowMs);
      this.#present(NOW, CLOCK, [{
        owner: NOW_OWNER,
        timestamp: nowMs,
        atom: offer(NOW, CLOCK, exact(encoded)),
      }]);
    }
    const shippedRows: Row[] = [];
    for (const owner of shipped) {
      shippedRows.push({
        owner: SHIPPED_OWNER,
        timestamp: 0n,
        atom: offer(SHIPPED, WIRE, exact(owner)),
      });
    }
    this.#present(SHIPPED, WIRE, shippedRows);
    const steps = Math.min(this.pending, Math.max(0, Math.floor(bound)));
    for (let step = 0; step < steps; step += 1) this.#step(this.#pop());
  }

  run(limit = 100_000): void {
    for (let turn = 0; turn < limit && this.pending !== 0; turn += 1) {
      this.turn(undefined, [], 64);
    }
    if (this.pending !== 0) throw new Error("no quiescence");
  }

  get pending(): number {
    return this.#frontier.length - this.#head;
  }

  hasFact(owner: Uint8Array): boolean {
    return this.#facts.has(key(owner));
  }

  hasDurable(owner: Uint8Array): boolean {
    return this.#durable.has(key(owner));
  }

  verdict(owner: Uint8Array): Verdict | undefined {
    return this.#memo.get(key(owner));
  }

  durable(owner: Uint8Array): Bytes | undefined {
    const data = this.#durable.get(key(owner));
    return data === undefined ? undefined : Buffer.from(data);
  }

  #step(owner: Bytes): void {
    const resident = this.#facts.get(key(owner));
    if (resident === undefined) return;
    const needs = resident.fact.atoms
      .filter((atom): atom is NeedAtom => atom.kind === Kind.Need)
      .map((atom) => materialize(atom, owner));
    let output = outcome(Verdict.Parked);
    if (needs.some((atom) => atom.effect === Effect.Suppress && this.validOffers(atom).length !== 0)) {
      output = outcome(Verdict.Suppressed);
    } else if (needs.some((atom) => atom.effect === Effect.Require && this.validOffers(atom).length === 0)) {
      output = outcome(Verdict.Parked);
    } else {
      const context: Answer[] = needs
        .filter((atom) => atom.effect === Effect.Require || atom.effect === Effect.Watch)
        .map((atom) => ({ need: atom, rows: this.validOffers(atom) }));
      output = this.#root.project(cloneFact(resident.fact), context) ?? outcome(Verdict.Parked);
    }
    this.#settle(owner, resident.fact, output);
  }

  #settle(owner: Bytes, fact: Fact, output: Out): void {
    const id = key(owner);
    this.#memo.set(id, output.verdict);
    const old = this.#owned.get(id) ?? [];
    this.#owned.delete(id);
    for (const row of old) this.#clean.get(addressKey(row.atom.role, row.atom.scope))?.remove(row);
    const fresh: Row[] = [];
    if (output.verdict === Verdict.Valid) {
      for (const atom of output.offers) {
        const row: Row = { owner, timestamp: 0n, atom: materialize(atom, owner) };
        this.#bucket(this.#clean, addressKey(atom.role, atom.scope)).add(row);
        fresh.push(row);
      }
    }
    if (fresh.length !== 0) this.#owned.set(id, fresh);

    // Duplicates remain visible in indexes, but wake-up deltas compare semantic row sets.
    const before = new Map(old.map((row) => [rowKey(row), row]));
    const after = new Map(fresh.map((row) => [rowKey(row), row]));
    for (const [rowId, row] of before) if (!after.has(rowId)) this.#wake(row.atom as OfferAtom, id);
    for (const [rowId, row] of after) if (!before.has(rowId)) this.#wake(row.atom as OfferAtom, id);
    if (output.verdict === Verdict.Reap || output.verdict === Verdict.Suppressed) this.#evict(owner, fact);
  }

  #evict(owner: Bytes, fact: Fact): void {
    const id = key(owner);
    this.#facts.delete(id);
    this.#memo.delete(id);
    this.#owned.delete(id);
    this.#durable.delete(id);
    for (const atom of fact.atoms) {
      this.#rows.get(indexKey(atom.kind, atom.role, atom.scope))?.remove({
        owner,
        timestamp: 0n,
        atom: materialize(atom, owner),
      });
    }
  }

  #present(role: Bytes, scope: Bytes, rows: Row[]): void {
    const bucket = new Bucket();
    for (const row of rows) bucket.add(row);
    this.#clean.set(addressKey(role, scope), bucket);
    for (const row of rows) this.#wake(row.atom as OfferAtom);
  }

  #wake(atom: OfferAtom, skip?: string): void {
    for (const row of this.needsFor(atom)) if (key(row.owner) !== skip) this.#enqueue(row.owner);
  }

  #enqueue(owner: Bytes): void {
    const id = key(owner);
    if (this.#queued.has(id)) return;
    this.#frontier.push(Buffer.from(owner));
    this.#queued.add(id);
  }

  #pop(): Bytes {
    const owner = this.#frontier[this.#head]!;
    this.#head += 1;
    this.#queued.delete(key(owner));
    if (this.#head === this.#frontier.length) {
      this.#frontier = [];
      this.#head = 0;
    } else if (this.#head > 1024 && this.#head * 2 >= this.#frontier.length) {
      this.#frontier = this.#frontier.slice(this.#head);
      this.#head = 0;
    }
    return owner;
  }

  #bucket(index: Map<string, Bucket>, id: string): Bucket {
    let bucket = index.get(id);
    if (bucket === undefined) {
      bucket = new Bucket();
      index.set(id, bucket);
    }
    return bucket;
  }

  #matching(bucket: Bucket | undefined, target: Target): Row[] {
    return bucket?.matching(target) ?? [];
  }
}

File: language-lab/go/kernel.go
package lab

import (
	"encoding/binary"
	"errors"
)

type Row struct {
	Owner     ID
	Timestamp uint64
	Atom      Atom
}

type Verdict uint8

const (
	// Valid is the zero value so a projector's Out{} means "valid, no offers",
	// matching the callback contract. Unknown is only seeded by admission.
	Valid Verdict = iota
	Invalid
	Parked
	Suppressed
	Reap
	Unknown
)

type Out struct {
	Verdict Verdict
	Offers  []Atom
}

func ValidOut(offers ...Atom) Out { return Out{Verdict: Valid, Offers: offers} }

type Answer struct {
	Need Atom
	Rows []Row
}

type Context []Answer

func By(context Context, role Blob) []Row {
	var rows []Row
	for _, answer := range context {
		if answer.Need.Role == role {
			rows = append(rows, answer.Rows...)
		}
	}
	return rows
}

// Root returns ok=false from Project when no family owns a fact tag.
type Root interface {
	Extract(Fact) bool
	Project(Fact, Context) (output Out, ok bool)
}

type Bucket struct {
	exact  map[Blob][]Row
	ranges []Row
}

func NewBucket() *Bucket { return &Bucket{exact: make(map[Blob][]Row)} }

func (bucket *Bucket) Add(row Row) {
	target := row.Atom.Target
	if target.Lo == target.Hi {
		bucket.exact[target.Lo] = append(bucket.exact[target.Lo], row)
		return
	}
	bucket.ranges = append(bucket.ranges, row)
}

func (bucket *Bucket) Remove(row Row) {
	target := row.Atom.Target
	if target.Lo == target.Hi {
		rows := bucket.exact[target.Lo]
		for i := range rows {
			if rows[i] == row {
				rows = append(rows[:i], rows[i+1:]...)
				break
			}
		}
		if len(rows) == 0 {
			delete(bucket.exact, target.Lo)
		} else {
			bucket.exact[target.Lo] = rows
		}
		return
	}
	for i := range bucket.ranges {
		if bucket.ranges[i] == row {
			bucket.ranges = append(bucket.ranges[:i], bucket.ranges[i+1:]...)
			return
		}
	}
}

func (bucket *Bucket) Matching(target Target) []Row {
	if target.Lo == target.Hi {
		rows := append([]Row(nil), bucket.exact[target.Lo]...)
		for _, row := range bucket.ranges {
			if row.Atom.Target.Lo <= target.Lo && target.Lo <= row.Atom.Target.Hi {
				rows = append(rows, row)
			}
		}
		return rows
	}
	var rows []Row
	for point, matches := range bucket.exact {
		if target.Lo <= point && point <= target.Hi {
			rows = append(rows, matches...)
		}
	}
	return rows
}

func (bucket *Bucket) All() []Row {
	rows := make([]Row, 0, bucket.Len())
	for _, matches := range bucket.exact {
		rows = append(rows, matches...)
	}
	return append(rows, bucket.ranges...)
}

func (bucket *Bucket) Len() int {
	count := len(bucket.ranges)
	for _, rows := range bucket.exact {
		count += len(rows)
	}
	return count
}

type rowKey struct {
	Kind        Kind
	Role, Scope Blob
}

type address struct{ Role, Scope Blob }

var (
	nowRole      = B("now")
	nowScope     = B("clock")
	shippedRole  = B("shipped")
	shippedScope = B("wire")
	nowOwner     = ID{0xff, 'n', 'o', 'w'}
	shippedOwner = ID{0xff, 's', 'h', 'i', 'p'}
)

type Node struct {
	root Root

	durable map[ID]Blob
	facts   map[ID]Fact
	rows    map[rowKey]*Bucket
	memo    map[ID]Verdict
	clean   map[address]*Bucket
	owned   map[ID][]Row

	frontier []ID
	head     int
	queued   map[ID]struct{}
}

func NewNode(root Root) *Node {
	return &Node{
		root:    root,
		durable: make(map[ID]Blob),
		facts:   make(map[ID]Fact),
		rows:    make(map[rowKey]*Bucket),
		memo:    make(map[ID]Verdict),
		clean:   make(map[address]*Bucket),
		owned:   make(map[ID][]Row),
		queued:  make(map[ID]struct{}),
	}
}

func (node *Node) Admit(data []byte) (ID, bool) {
	fact, err := Decode(data)
	if err != nil {
		return ID{}, false
	}
	owner, err := FactID(fact)
	if err != nil {
		return ID{}, false
	}
	if _, exists := node.facts[owner]; exists {
		return owner, true
	}
	node.facts[owner] = fact
	node.memo[owner] = Unknown
	if node.root.Extract(cloneFact(fact)) {
		node.durable[owner] = Blob(string(data))
	}
	for _, atom := range fact.Atoms {
		key := rowKey{Kind: atom.Kind, Role: atom.Role, Scope: atom.Scope}
		bucket := node.rows[key]
		if bucket == nil {
			bucket = NewBucket()
			node.rows[key] = bucket
		}
		bucket.Add(Row{Owner: owner, Atom: Materialize(atom, owner)})
	}
	node.enqueue(owner)
	return owner, true
}

func (node *Node) OffersFor(need Atom) []Row {
	return matching(node.rows[rowKey{Kind: Offer, Role: need.Role, Scope: need.Scope}], need.Target)
}

func (node *Node) NeedsFor(offer Atom) []Row {
	return matching(node.rows[rowKey{Kind: Need, Role: offer.Role, Scope: offer.Scope}], offer.Target)
}

func (node *Node) ValidOffers(need Atom) []Row {
	return matching(node.clean[address{Role: need.Role, Scope: need.Scope}], need.Target)
}

func (node *Node) Watched(role, scope Blob) []Row {
	bucket := node.clean[address{Role: role, Scope: scope}]
	if bucket == nil {
		return nil
	}
	return bucket.All()
}

func (node *Node) Turn(nowMS *uint64, shipped []ID, bound int) {
	if nowMS != nil {
		var encoded [8]byte
		binary.BigEndian.PutUint64(encoded[:], *nowMS)
		now := Blob(string(encoded[:]))
		node.present(nowRole, nowScope, []Row{{
			Owner: nowOwner, Timestamp: *nowMS,
			Atom: OfferAtom(nowRole, nowScope, Exact(now)),
		}})
	}
	shippedRows := make([]Row, 0, len(shipped))
	for _, owner := range shipped {
		shippedRows = append(shippedRows, Row{
			Owner: shippedOwner,
			Atom:  OfferAtom(shippedRole, shippedScope, Exact(owner.Blob())),
		})
	}
	node.present(shippedRole, shippedScope, shippedRows)
	steps := node.Pending()
	if bound < steps {
		steps = bound
	}
	if steps < 0 {
		steps = 0
	}
	for range steps {
		owner := node.pop()
		node.step(owner)
	}
}

func (node *Node) Run() error {
	for range 100_000 {
		if node.Pending() == 0 {
			return nil
		}
		node.Turn(nil, nil, 64)
	}
	return errors.New("no quiescence")
}

func (node *Node) Pending() int { return len(node.frontier) - node.head }

func (node *Node) HasFact(owner ID) bool {
	_, ok := node.facts[owner]
	return ok
}

func (node *Node) HasDurable(owner ID) bool {
	_, ok := node.durable[owner]
	return ok
}

func (node *Node) Verdict(owner ID) (Verdict, bool) {
	verdict, ok := node.memo[owner]
	return verdict, ok
}

func (node *Node) Durable(owner ID) ([]byte, bool) {
	data, ok := node.durable[owner]
	return blobBytes(data), ok
}

func (node *Node) step(owner ID) {
	fact, exists := node.facts[owner]
	if !exists {
		return
	}
	needs := make([]Atom, 0)
	for _, atom := range fact.Atoms {
		if atom.Kind == Need {
			needs = append(needs, Materialize(atom, owner))
		}
	}
	output := Out{Verdict: Parked}
	if anyNeed(needs, Suppress, func(need Atom) bool { return len(node.ValidOffers(need)) != 0 }) {
		output = Out{Verdict: Suppressed}
	} else if anyNeed(needs, Require, func(need Atom) bool { return len(node.ValidOffers(need)) == 0 }) {
		output = Out{Verdict: Parked}
	} else {
		context := make(Context, 0, len(needs))
		for _, need := range needs {
			if need.Effect == Require || need.Effect == Watch {
				context = append(context, Answer{Need: need, Rows: node.ValidOffers(need)})
			}
		}
		if projected, ok := node.root.Project(cloneFact(fact), context); ok {
			output = projected
		}
	}
	node.settle(owner, fact, output)
}

func anyNeed(needs []Atom, effect Effect, predicate func(Atom) bool) bool {
	for _, need := range needs {
		if need.Effect == effect && predicate(need) {
			return true
		}
	}
	return false
}

func (node *Node) settle(owner ID, fact Fact, output Out) {
	node.memo[owner] = output.Verdict
	old := node.owned[owner]
	delete(node.owned, owner)
	for _, row := range old {
		node.clean[address{Role: row.Atom.Role, Scope: row.Atom.Scope}].Remove(row)
	}
	var fresh []Row
	if output.Verdict == Valid {
		fresh = make([]Row, 0, len(output.Offers))
		for _, atom := range output.Offers {
			row := Row{Owner: owner, Atom: Materialize(atom, owner)}
			key := address{Role: row.Atom.Role, Scope: row.Atom.Scope}
			bucket := node.clean[key]
			if bucket == nil {
				bucket = NewBucket()
				node.clean[key] = bucket
			}
			bucket.Add(row)
			fresh = append(fresh, row)
		}
	}
	if len(fresh) != 0 {
		node.owned[owner] = fresh
	}
	for row := range changedRows(old, fresh) {
		node.wake(row.Atom, &owner)
	}
	if output.Verdict == Reap || output.Verdict == Suppressed {
		node.evict(owner, fact)
	}
}

func changedRows(old, fresh []Row) map[Row]struct{} {
	changed := make(map[Row]struct{}, len(old)+len(fresh))
	for _, row := range old {
		changed[row] = struct{}{}
	}
	for _, row := range fresh {
		if _, exists := changed[row]; exists {
			delete(changed, row)
		} else {
			changed[row] = struct{}{}
		}
	}
	return changed
}

func (node *Node) evict(owner ID, fact Fact) {
	delete(node.facts, owner)
	delete(node.memo, owner)
	delete(node.owned, owner)
	delete(node.durable, owner)
	for _, atom := range fact.Atoms {
		key := rowKey{Kind: atom.Kind, Role: atom.Role, Scope: atom.Scope}
		if bucket := node.rows[key]; bucket != nil {
			bucket.Remove(Row{Owner: owner, Atom: Materialize(atom, owner)})
		}
	}
}

func (node *Node) present(role, scope Blob, rows []Row) {
	bucket := NewBucket()
	for _, row := range rows {
		bucket.Add(row)
	}
	node.clean[address{Role: role, Scope: scope}] = bucket
	for _, row := range rows {
		node.wake(row.Atom, nil)
	}
}

func (node *Node) wake(offer Atom, skip *ID) {
	for _, row := range node.NeedsFor(offer) {
		if skip == nil || row.Owner != *skip {
			node.enqueue(row.Owner)
		}
	}
}

func (node *Node) enqueue(owner ID) {
	if _, exists := node.queued[owner]; exists {
		return
	}
	node.frontier = append(node.frontier, owner)
	node.queued[owner] = struct{}{}
}

func (node *Node) pop() ID {
	owner := node.frontier[node.head]
	node.frontier[node.head] = ID{}
	node.head++
	delete(node.queued, owner)
	if node.head == len(node.frontier) {
		node.frontier = nil
		node.head = 0
	} else if node.head > 1024 && node.head*2 >= len(node.frontier) {
		node.frontier = append([]ID(nil), node.frontier[node.head:]...)
		node.head = 0
	}
	return owner
}

func matching(bucket *Bucket, target Target) []Row {
	if bucket == nil {
		return nil
	}
	return bucket.Matching(target)
}

func cloneFact(fact Fact) Fact {
	return Fact{Tag: fact.Tag, Atoms: append([]Atom(nil), fact.Atoms...)}
}
